Prose/ProseLib/Sources/App/AppReducer.swift

```diff
@@ -20,37 +20,45 @@ public let appReducer: Reducer<
     AppAction,
     AppEnvironment
 > = Reducer.combine([
-    authenticationReducer._pullback(
-        state: (\AppState.route).case(/AppRoute.auth),
+    authenticationReducer.optional().pullback(
+        state: \AppState.auth,
         action: /AppAction.auth,
-        environment: \AppEnvironment.auth
+        environment: { $0.auth }
     ),
-    mainWindowReducer._pullback(
-        state: (\AppState.route).case(/AppRoute.main),
+    mainWindowReducer.optional().pullback(
+        state: \AppState.main,
         action: /AppAction.main,
-        environment: \AppEnvironment.main
+        environment: { $0.main }
     ),
     Reducer { state, action, environment in
         switch action {
         case .onAppear:
-            guard case .auth = state.route else { return .none }
+            guard !state.hasAppearedAtLeastOnce else { return .none }
+            state.hasAppearedAtLeastOnce = true
 
-            guard let jid = environment.userDefaults.loadCurrentAccount() else { return .none }
-
-            do {
-                guard let password = try environment.credentials.loadCredentials(jid) else { return .none }
-                return Effect(value: .auth(.didLogIn(jid: jid, password: password)))
-            } catch {
-                // User might have denied access to the credentials
-                print("Error when loading credentials: \(error.localizedDescription)")
-
-                // TODO: [Rémi Bardon] Add an optional error message to the login screen,
-                //       to show why the login screen appears in case of errors like this.
-                state.route = .auth(.basicAuth(BasicAuthState(jid: jid.jidString)))
+            if let jid = environment.userDefaults.loadCurrentAccount() {
```

Comment:
Since this routine grew in complexity once more, now I'd say let's convert this into a proper side-effect. Where `.onAppear` returns an effect with an output value of `AppAction.authenticationResponse(Result<Credentials, EquatableError>)`.

Reply:
I don't really get what you mean. This is a bit verbose, but I'm not sure we would benefit from adding a new `Action` (if that's what you meant). I already created `. requireAuthentication` to make the process easier to understand.

I could add more comments though.

Reply:
That would be my proposal…

I cannot say much other than that side-effects are (mostly) wrapped into actual `Effect`s, that it has less nesting levels and feels more idiomatic to me 🤷.

It also does away with the "An "optional" reducer at "App/AppReducer.swift:40" received an action when state was "nil" message because the reducer doesn't send `.auth(.didLogIn())` itself anymore.

I've also added a `Credentials` struct which has the fields `jid` and `password`.

```swift
public let appReducer: Reducer<
    AppState,
    AppAction,
    AppEnvironment
> = Reducer.combine([
    authenticationReducer.optional().pullback(
        state: \AppState.auth,
        action: /AppAction.auth,
        environment: { $0.auth }
    ),
    mainWindowReducer.optional().pullback(
        state: \AppState.main,
        action: /AppAction.main,
        environment: { $0.main }
    ),
    Reducer { state, action, environment in
        func proceedToMainFlow(with credentials: Credentials) {
            state.main = MainScreenState(
                sidebar: .init(
                    credentials: .init(jid: credentials.jid),
                    footer: .init(
                        avatar: .init(
                            // TODO: Use a JID type, to avoid this parsing
                            // TODO: Use an image stored by the user (not a preview asset)
                            avatar: "avatars/\(credentials.jid.node ?? "valerian")"
                        )
                    )
                )
            )
            state.auth = nil
        }
        
        func proceedToLogin() {
          state.auth = .init(route:
            .basicAuth(.init(jid: environment.userDefaults.loadCurrentAccount()?.rawValue ?? "")))
        }
    
        switch action {
        case .onAppear where !state.hasAppearedAtLeastOnce:
          return Effect<Credentials?, EquatableError>.result {
              Result {
                try environment.userDefaults.loadCurrentAccount()
                  .flatMap(environment.credentials.loadCredentials)
              }.mapError(EquatableError.init)
          }
          .catchToEffect()
          .map(AppAction.authenticationResult)
        
        case let .authenticationResult(.success(.some(credentials))):
            proceedToMainFlow(with: credentials)
            return .none
        
        case .authenticationResult(.success(.none)):
          proceedToLogin()
          return .none
        
        case let .authenticationResult(.failure(error)):
          print("Error when loading credentials: \(error.localizedDescription)")
          proceedToLogin()
          return .none
        
        case let .auth(.didLogIn(credentials)):
          proceedToMainFlow(with: credentials)
          return .fireAndForget {
              environment.userDefaults.saveCurrentAccount(credentials.jid)
              try? environment.credentials.save(credentials.jid, credentials.password)
          }
        
        case .onAppear, .auth, .main:
          return .none
    }
  }
])
```